### PR TITLE
ElasticsearchWriter: don't leak sockets

### LINKED ISSUE #7018: ElasticsearchWriter not closing SSL connections on  Icinga2 2.10.3.1
## Expected Behavior

I would expect Icinga2 to close connections to Elasticsearch rather than keeping them open and opening additional ones.

## Current Behavior

This issue is as the same as #6989 but for ElasticsearchWriter with SSL connections. For non-SSL connections, Icinga2 closes connection as normal.
## Possible Solution

I think to solution would be the same as #6990 but I'm not sure about that.
## Steps to Reproduce (for bugs)

1. Turn on elasticsearch feature with SSL configuration and monitor established TCP connections. The connection count increases after each data flush (for example 10 seconds in my case)
2. Disable SSL in configuration. The connection count does not increase.

## Context

This issue is causing Icinga2 to run out of file descriptor and every check will fail. 
## Your Environment

* Version used (`icinga2 --version`): 2.10.3-1
* Operating System and version: Ubuntu 16.04
* Enabled features (`icinga2 feature list`): api checker elasticsearch ido-mysql influxdb mainlog syslog
* Icinga Web 2 version and modules (System - About): Icinga Web 2 Version 2.6.2
* Config validation (`icinga2 daemon -C`):
[2019-03-15 15:27:33 +0700] information/cli: Icinga application loader (version: r2.10.3-1)
[2019-03-15 15:27:33 +0700] information/cli: Loading configuration file(s).
[2019-03-15 15:27:34 +0700] information/ConfigItem: Committing config item(s).
[2019-03-15 15:27:34 +0700] information/ApiListener: My API identity: REMOVED
.....
[2019-03-15 15:27:35 +0700] information/ScriptGlobal: Dumping variables to file '/var/cache/icinga2/icinga2.vars'
[2019-03-15 15:27:35 +0700] information/cli: Finished validating the configuration file(s).

* If you run multiple Icinga 2 instances, the `zones.conf` file (or `icinga2 object list --type Endpoint` and `icinga2 object list --type Zone`) from all affected nodes.
* ElasticseachWriter configuration: /etc/icinga2/features-enabled/elasticsearch.conf
object ElasticsearchWriter "elasticsearch" {
  host = "127.0.0.1"
  port = 9200
  index = "icinga2_testbed"
  enable_tls = true
  username = "icinga2"
  password = "REMOVED"
  enable_send_perfdata = true
  flush_threshold = 1024
  flush_interval = 10s
}

* Elasticsearch:  elasticsearch-6.6.2-1

### PR DESCRIPTION
**Pull request by Al2Klimov**
fixes #7018

### CODE REVIEW

**Comment by Al2Klimov** [comment]
## TODO

* [ ] #6990

File: lib/base/defer.hpp [hunk 1]
```diff
@@ -30,7 +30,11 @@ class Defer
 	inline
 	~Defer()
 	{
-		m_Func();
+		try {
+			m_Func();
+		} catch (...) {
+			// https://stackoverflow.com/questions/130117/throwing-exceptions-out-of-a-destructor
+		}
```
Comment (dnsmichi) [inline lib/base/defer.hpp]: Please split this into a separate PR.

Comment (Al2Klimov) [reply lib/base/defer.hpp]: The first commit will disappear automatically once the PR linked above has been merged.

**Comment by dnsmichi** [comment]
@Al2Klimov please rebase.

**Comment by dnsmichi** [comment]
Tests: https://github.com/Icinga/icinga2/issues/7018#issuecomment-473964773

**Pull request merged**